apps/showcase/src/components/training/code-editor-view/code-editor-view.component.ts

```diff
@@ -0,0 +1,199 @@
+import {FileSystemTree} from '@webcontainer/api';
+import {MonacoEditorModule} from 'ngx-monaco-editor-v2';
+import {MonacoTreeElement, NgxMonacoTreeComponent} from 'ngx-monaco-tree';
+import {
+  combineLatestWith,
+  debounceTime,
+  distinctUntilChanged,
+  filter,
+  from,
+  map,
+  of,
+  skip,
+  startWith,
+  switchMap
+} from 'rxjs';
+import {WebContainerService} from '../../../services';
+import {CodeEditorControlComponent} from '../code-editor-control';
+import {takeUntilDestroyed} from '@angular/core/rxjs-interop';
+
+/** ngx-monaco-editor options language - determined based on file extension */
+const editorOptionsLanguage: Record<string, string> = {
+  html: 'xml',
+  json: 'json',
+  ts: 'typescript',
+  js: 'javascript'
+};
+
+/** Mode of Code Editor */
+export type EditorMode = 'readonly' | 'interactive';
+
+/** Project properties for the current exercise */
+export interface TrainingProject {
+  /** Starting file to be displayed in the project */
+  startingFile: string;
+  /** Files in the project */
+  files: FileSystemTree;
+  /** Commands to run in the project */
+  commands: string[];
+  /** Current working directory in project */
+  cwd: string;
+}
+
+@Component({
+  selector: 'code-editor-view',
+  standalone: true,
+  imports: [
+    AsyncPipe,
+    FormsModule,
+    MonacoEditorModule,
+    ReactiveFormsModule,
+    NgxMonacoTreeComponent,
+    CodeEditorControlComponent
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+  templateUrl: './code-editor-view.component.html',
+  styleUrl: './code-editor-view.component.scss'
+})
+export class CodeEditorViewComponent implements OnDestroy, OnChanges {
+  /**
+   * Allow to edit the code in the monaco editor
+   */
+  @Input() public editorMode: EditorMode = 'readonly';
+  /**
+   * Project to load in the code editor.
+   * It should describe the files to load, the starting file, the folder dedicated to the project as well as the
+   * commands to initialize the project
+   */
+  @Input() public project?: TrainingProject;
+
+  /**
+   * Service to load files and run commands in the application instance of the webcontainer.
+   */
+  public readonly webContainerService = inject(WebContainerService);
+  /**
+   * File tree loaded in the project folder within the web container instance.
+   */
+  public tree$ = this.webContainerService.monacoTree$;
+  /**
+   * @see {FormBuilder}
+   */
+  private readonly formBuilder = inject(FormBuilder);
+
+  /**
+   * Form with the selected file and its content which can be edited in the Monaco Editor
+   */
+  public form: FormGroup<{
+    code: FormControl<string | null>;
+    file: FormControl<string | null>;
+  }> = this.formBuilder.group({
+      code: '',
+      file: this.project?.startingFile || ''
+    });
+  /**
+   * Configuration for the Monaco Editor
+   */
+  public editorOptions$ = this.form.controls.file.valueChanges.pipe(
+    startWith(this.project?.startingFile || ''),
```

Comment:
`this.project` will always be undefined in constructor, right ?

Reply:
It is bound on the form value change event